ref/haskell/src/Codec/Binary/Bech32.hs

```diff
@@ -0,0 +1,128 @@
+module Codec.Binary.Bech32
+  ( bech32Encode
+  , bech32Decode
+  , segwitEncode
+  , segwitDecode
+  , Word5(UnsafeWord5)
+  ) where
+
+import Control.Monad (guard)
+import Data.Bits (Bits, unsafeShiftL, unsafeShiftR, (.&.), xor, testBit)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
+import Data.Char (toLower, toUpper)
+import Data.Digits
+import Data.Foldable (foldl')
+import qualified Data.HashMap.Strict as Map
+import Data.Word (Word8, Word32)
+import Data.Hashable (Hashable(..))
+
+type HRP = BS.ByteString
+type Data = [Word8]
+
+(.>>.), (.<<.) :: Bits a => a -> Int -> a
+(.>>.) = unsafeShiftR
+(.<<.) = unsafeShiftL
+
+newtype Word5 = UnsafeWord5 Word8
+              deriving (Eq)
+
+instance Hashable Word5 where
+    hashWithSalt salt (UnsafeWord5 x) = hashWithSalt salt x
+
+word5 :: Word8 -> Word5
+word5 x = UnsafeWord5 (x .&. 31)
+
+fromWord5 :: Num a => Word5 -> a
+fromWord5 (UnsafeWord5 x) = fromIntegral x
+
+charset :: String
+charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
```

Comment:
The hash-map uses `IntMap` as it's implementation which is a Patricia trie. To look up an item in a Patrica trie will requires traversing several indirections.  On top of that there is going to be quite a bit of overhead in the Patrica trie structure itself.  Even with the 11 or so empty cells in the array representation, the memory use is probably going to be less than the Patrica trie.  The array lookup is just an offset and and at most a single indirection after that.  Plus we get the very minor bonus of dropping one popular, but large dependency since Arrays are part of the language standard.

If you are worried about complexity, you can add an exhaustive test case in your set of tests:

```
charsetMap_test :: Bool
charsetMap_test = all correct1 [char i | i <- [0..255]] && all correct2 [word5 i | i <- [0..31]]
  where
    correct1 c = fromMaybe c (charsetMapInv <$> charsetMap c) = c
    correct2 w = charsetMap (charsetMapInv w) = Just w
```

That said, if you really really want `HashMap` I'll concede.